src/state.rs

```diff
@@ -328,7 +331,19 @@ impl VMState {
             Predicate::Le => self.flag_eq || self.flag_lt_of,
             Predicate::Ne => !self.flag_eq,
             Predicate::GtOrLt => self.flag_gt || self.flag_lt_of,
+        };
+        let mut kernel_ok = true;
```

Comment:
I would do `kernel_ok =  if opcode.variant.requires_kernel_mode() { self.current_context()?.is_kernel() } else { true }`. But it is just a biased opinion, you don't have to change it. Same applies for the context